File: ios/Sources/RynBridgePayment/PaymentModule.swift
import Foundation
import RynBridge

public struct PaymentModule: BridgeModule, Sendable {
    public let name = "payment"
    public let version = "0.1.0"
    public let actions: [String: ActionHandler]

    public init(provider: PaymentProvider) {
        actions = [
            "getProducts": { payload in
                guard let ids = payload["productIds"]?.arrayValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: productIds")
                }
                let productIds = ids.compactMap { $0.stringValue }
                let products = try await provider.getProducts(productIds: productIds)
                return ["products": .array(products.map { .dictionary($0.toPayload()) })]
            },
            "purchase": { payload in
                guard let productId = payload["productId"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: productId")
                }
                let quantity = payload["quantity"]?.intValue ?? 1
                let result = try await provider.purchase(productId: productId, quantity: quantity)
                return result.toPayload()
            },
            "restorePurchases": { _ in
                let transactions = try await provider.restorePurchases()
                return ["transactions": .array(transactions.map { .dictionary($0.toPayload()) })]
            },
            "finishTransaction": { payload in
                guard let transactionId = payload["transactionId"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: transactionId")
                }
                try await provider.finishTransaction(transactionId: transactionId)
                return [:]
            },
        ]
    }
}

File: android/payment/src/main/kotlin/io/rynbridge/payment/PaymentModule.kt
package io.rynbridge.payment

import io.rynbridge.core.*

class PaymentModule(provider: PaymentProvider) : BridgeModule {

    override val name = "payment"
    override val version = "0.1.0"
    override val actions: Map<String, ActionHandler> = mapOf(
        "getProducts" to { payload ->
            val ids = payload["productIds"]?.arrayValue
                ?.mapNotNull { it.stringValue }
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: productIds")
            val products = provider.getProducts(ids)
            mapOf("products" to BridgeValue.array(products.map { BridgeValue.obj(it.toPayload()) }))
        },
        "purchase" to { payload ->
            val productId = payload["productId"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: productId")
            val quantity = payload["quantity"]?.intValue?.toInt() ?: 1
            val result = provider.purchase(productId, quantity)
            result.toPayload()
        },
        "restorePurchases" to { _ ->
            val transactions = provider.restorePurchases()
            mapOf("transactions" to BridgeValue.array(transactions.map { BridgeValue.obj(it.toPayload()) }))
        },
        "finishTransaction" to { payload ->
            val transactionId = payload["transactionId"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: transactionId")
            provider.finishTransaction(transactionId)
            emptyMap()
        }
    )
}
